src/Interpreters/TreeOptimizer.cpp

```diff
@@ -587,6 +589,56 @@ void optimizeFunctionsToSubcolumns(ASTPtr & query, const StorageMetadataPtr & me
     RewriteFunctionToSubcolumnVisitor(data).visit(query);
 }
 
+/// Rewrites multi quantile()() functions with the same arguments to quantiles()()[]
+/// eg:SELECT quantile(0.5)(x), quantile(0.9)(x), quantile(0.95)(x) FROM...
+///    rewrite to : SELECT quantiles(0.5, 0.9, 0.95)(x)[1], quantiles(0.5, 0.9, 0.95)(x)[2], quantiles(0.5, 0.9, 0.95)(x)[3] FROM ...
+void fuseCandidate(std::unordered_map<String, GatherFunctionQuantileData::FuseQuantileAggregatesData> & fuse_quantile)
+{
+    for (auto candidate : fuse_quantile)
+    {
+        String func_name = candidate.first;
+        GatherFunctionQuantileData::FuseQuantileAggregatesData args_to_functions = candidate.second;
+
+        // Try to fuse multiply quantilexxx Function to one
+        for (auto it : args_to_functions.arg_map_function)
+        {
+            std::vector<ASTFunction *> functions = it.second;
+            size_t count = functions.size();
+            if (count > 1)
+            {
+                auto param_exp_list = std::make_shared<ASTExpressionList>();
+                for (auto func : functions)
+                {
+                    const ASTs & parameters = func->parameters->as<ASTExpressionList &>().children;
+                    assert(parameters.size() == 1);
+                    param_exp_list->children.push_back(parameters[0]);
+                }
+                functions[0]->parameters = param_exp_list;
+                functions[0]->name = quantile_fuse_name_mapping.find(func_name)->second;
+                auto func_base = functions[0]->clone();
+
+                for (size_t i = 0; i < count; ++i)
+                {
+                    functions[i]->name = "arrayElement";
+                    auto func_exp_list = std::make_shared<ASTExpressionList>();
+                    func_exp_list->children.push_back(func_base);
+                    func_exp_list->children.push_back(std::make_shared<ASTLiteral>(i + 1));
+                    functions[i]->children.clear();
+                    functions[i]->parameters = nullptr;
+                    functions[i]->arguments = func_exp_list;
+                    functions[i]->children.push_back(func_exp_list);
```

Comment:
Maybe `makeASTFunction` will be more convenient here? https://github.com/ClickHouse/ClickHouse/blob/817ed354ff6cde9e06d33e9749cbf65a54eeac35/src/Parsers/ASTFunction.h#L64-L76

Reply:
fixed.